### PR TITLE
⚡RoPE

### PR DESCRIPTION
**Pull request by lapp0**
Eliminates virtually all RoPE overhead.

### Numerically equivalent
- [x] Shared `cos` / `sin` buffer for video / audio
- [x] No expensive `stack()` operation in `forward()`

### Not numerically equivalent
These changes retain q/k dot-product invariance of RoPE with the same level of precision as upstream, but change behavior otherwise.

- [x] Symmetrize rotations
- [x] Downsampling of sin / cos buffers (rotation has equivalent precision w/ downsampling iff we symmetrize)

# TODO
- [x] 1x4090 benchmark / profile
- [ ] 1xH100 benchmark / profile
- [ ] Measure loss in real train run to measure impact of symmetrization and downsampling (**need a checkpoint for this**)

### CODE REVIEW

File: owl_wms/nn/attn.py [hunk 1]
```diff
@@ -70,16 +70,20 @@ def forward(self, x, block_mask, kv_cache=None):
...
             old_k, old_v = kv_cache.get(self.layer_ind)
             k = torch.cat([old_k, k], dim=2)
             v = torch.cat([old_v, v], dim=2)
-            if kv_cache.should_update:
-                kv_cache.update(k.clone(), v.clone(), self.layer_ind)
 
-        q, k = self.rope(q, offset=offset), self.rope(k)
+        # update cache
+        if kv_cache is not None and kv_cache.should_update:
+            kv_cache.update(k.clone(), v.clone(), self.layer_ind)
```
Comment (lapp0) [inline owl_wms/nn/attn.py:86]: unnecessary clone, can shave time here.
